internal/sourcemap/metadata_fetcher.go

```diff
@@ -0,0 +1,368 @@
+	init             chan struct{}
+	initErr          error
+	invalidationChan chan<- []identifier
+}
+
+func NewMetadataFetcher(ctx context.Context, esClient *elasticsearch.Client, index string) (MetadataFetcher, <-chan []identifier) {
+	invalidationCh := make(chan []identifier)
+
+	s := &MetadataESFetcher{
+		esClient:         esClient,
+		index:            index,
+		set:              make(map[identifier]string),
+		alias:            make(map[identifier]*identifier),
+		logger:           logp.NewLogger(logs.Sourcemap),
+		init:             make(chan struct{}),
+		invalidationChan: invalidationCh,
+	}
+
+	s.startBackgroundSync(ctx)
+
+	return s, invalidationCh
+}
+
+func (s *MetadataESFetcher) getID(key identifier) (*identifier, bool) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	if _, ok := s.set[key]; ok {
+		return &key, ok
+	}
+
+	// path is missing from the metadata cache (and ES).
+	// Is it an alias ?
+	// Try to retrieve the sourcemap from the alias map
+	i, ok := s.alias[key]
+	return i, ok
+}
+
+func (s *MetadataESFetcher) ready() <-chan struct{} {
+	return s.init
+}
+
+func (s *MetadataESFetcher) err() error {
+	select {
+	case <-s.ready():
+		s.mu.RLock()
+		defer s.mu.RUnlock()
+		return s.initErr
+	default:
+		return errors.New("metadata es fetcher not ready")
+	}
+}
+
+func (s *MetadataESFetcher) startBackgroundSync(parent context.Context) {
+	go func() {
+		s.logger.Debug("populating metadata cache")
+
+		ctx, cancel := context.WithTimeout(parent, 1*time.Second)
+		err := s.ping(ctx)
+		cancel()
+
+		if err != nil {
+			// it is fine to not lock here since err will not access
+			// initErr until the init channel is closed.
+			s.initErr = fmt.Errorf("failed to ping es cluster: %w: %v", errFetcherUnvailable, err)
+			s.logger.Error(s.initErr)
+		} else {
+			// First run, populate cache
+			ctx, cancel = context.WithTimeout(parent, syncTimeout)
+			err := s.sync(ctx)
+			cancel()
+
+			s.initErr = err
+
+			if err != nil {
+				s.logger.Errorf("failed to fetch sourcemaps metadata: %v", err)
+			} else {
+				// only close the init chan and mark the fetcher as ready if
+				// sync succeeded
+				s.logger.Info("init routine completed")
+			}
+		}
+
+		close(s.init)
+
+		// TODO make this a config option ?
+		t := time.NewTicker(30 * time.Second)
+		defer t.Stop()
+
+		for {
+			select {
+			case <-t.C:
+				ctx, cancel := context.WithTimeout(parent, syncTimeout)
+
+				if err := s.sync(ctx); err != nil {
+					s.logger.Errorf("failed to sync sourcemaps metadata: %v", err)
+				}
+
+				cancel()
+			case <-parent.Done():
+				s.logger.Info("update routine done")
+				// close invalidation channel
+				close(s.invalidationChan)
+				return
+			}
+		}
+	}()
+}
+
+func (s *MetadataESFetcher) ping(ctx context.Context) error {
+	// we cannot use PingRequest because the library is
+	// building a broken url and the request is timing out.
+	req := esapi.IndicesGetRequest{
+		Index: []string{s.index},
+	}
+	_, err := req.Do(ctx, s.esClient)
```

Comment:
We should be closing the response body when err == nil.

Reply:
:facepalm: